File: src/main/java/com/gl/module/system/vo/SysMenuVO.java
package com.gl.module.system.vo;

import org.hibernate.validator.constraints.NotBlank;

/**
 * @author gantrylau
 * @since 2017/2/21
 */
public class SysMenuVO {

    private Long id;

    @NotBlank
    private String name;

    @NotBlank
    private String alias;

    private Integer sort;

    private String parentAlias;

    private boolean leaf;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getAlias() {
        return alias;
    }

    public void setAlias(String alias) {
        this.alias = alias;
    }

    public Integer getSort() {
        return sort;
    }

    public void setSort(Integer sort) {
        this.sort = sort;
    }

    public String getParentAlias() {
        return parentAlias;
    }

    public void setParentAlias(String parentAlias) {
        this.parentAlias = parentAlias;
    }

    public boolean isLeaf() {
        return leaf;
    }

    public void setLeaf(boolean leaf) {
        this.leaf = leaf;
    }
}

File: src/main/java/com/gl/module/system/vo/SysMenuVO.kt
package com.gl.module.system.vo

import com.gl.module.system.domain.SysMenu
import java.util.*


/**
 * @author gantrylau
 * *
 * @since 2017/2/17
 */
class SysMenuVO(menu: SysMenu) : Cloneable {
    var id: Long? = null
    var name: String? = null
    var alias: String? = null
    var sort: Int? = null
    var parentAlias: String? = null
    var children: MutableList<SysMenuVO>? = null
    var isLeaf: Boolean = false

    init {
        this.id = menu.id
        this.name = menu.name
        this.alias = menu.alias
        this.sort = menu.sort
        this.parentAlias = menu.parentAlias
        this.isLeaf = menu.isLeaf
    }

    fun addChild(list: List<SysMenuVO>) {
        if (this.children == null)
            this.children = mutableListOf()
        this.children?.addAll(list)
    }

    @Throws(CloneNotSupportedException::class)
    override fun clone(): Any {
        return super.clone()
    }

    companion object {

        fun toTree(menus: List<SysMenu>): List<SysMenuVO> {
            val list = mutableListOf<SysMenuVO>()
            menus.mapTo(list) { SysMenuVO(it) }
            list.sortedWith(Comparator { t1, t2 ->
                t1.sort?.compareTo(t2.sort ?: 0) ?: -1
            })
            return formatTree(list, null)
        }

        fun formatTree(list: List<SysMenuVO>, parentAlias: String?): List<SysMenuVO> {
            val result = mutableListOf<SysMenuVO>()
            for (vo in list) {
                if ((parentAlias == null && vo.parentAlias == null) || (parentAlias != null && parentAlias == vo.parentAlias)) {
                    if (!vo.isLeaf)
                        vo.addChild(formatTree(list, vo.alias))
                    result.add(vo)
                }
            }
            return result
        }
    }
}